### ISSUE TITLE
"tinyformat: Too many conversion specifiers in format string"

### ISSUE THREAD

**Issue by rebroad** [bitcoin/bitcoin#9423]
I don't think I've seen an issue raised for this - it's more of a development issue than a runtime issue but if encounter it often, which is that if there is an incorrect number of arguments in a LogPrint() or LogPrintf() then no compile error/warning is given, it will cause the code to SIGABRT, and a gdb backtrace never provides much useful information about where the error in the code is. Only the name of the thread is given which can often leave a lot of searching or git bisecting. This common problem could be made easier to debug...

**Comment by gmaxwell** [comment]
It's basically impossible to give a compile error for this unfortunately. It's only possible for printf because the compiler basically implements the whole function internally.

**Comment by ryanofsky** [comment]
Could \_\_attribute__ format work?

http://clang.llvm.org/docs/AttributeReference.html#format-gnu-format
https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html

**Comment by laanwj** [comment]
>  it will cause the code to SIGABRT, and a gdb backtrace never provides much useful information about where the error in the code is

That's not entirely correct. It is set up to raise an exception.

```
 #define TINYFORMAT_ERROR(reasonString) throw std::runtime_error(reasonString)
```

In the case of logging it would be OK to just log a warning message instead of raising an exception.
The various Log* could catch this exception and do that.

`strprintf` OTOH should not, because it is also used in more critical places than logging (such as formatting RPC output), and a programmer error there to specify not enough conversion specifiers is a critical problem.

**Comment by laanwj** [comment]
An external script would also be able to check this, without too much trouble as the format strings must be static.

In the long run I'd prefer to abandon strprintf-ish syntax and use, at least for logging, the more C++ish way of
```
Logger(Error,"rpc") << "HTTP: starting" << rpcThreads << "worker threads";
```
This is what tinyformat does internally. It is a type-safe compatibility layer from printf-like vararg syntax to c++ formatting.

Using tinyformat was a workaround to make it possible to keep most of the logging lines the same while still avoiding libc printf's pitfalls such as variability between platforms and type-safety issues. Lack of compile-time checking of number of arguments is annoying though.

**Comment by fanquake** [comment]
Fixed in #9423.
